shared/src/main/scala/fetch.scala

```diff
@@ -60,22 +60,30 @@ object Query {
 
 /** Requests in Fetch Free monad.
   */
-sealed trait FetchRequest[I, A] extends Product with Serializable {
+sealed trait FetchRequest extends Product with Serializable {
   def fullfilledBy(cache: DataSourceCache): Boolean
+}
+
+sealed trait FetchQuery[I, A] extends FetchRequest {
   def missingIdentities(cache: DataSourceCache): List[I]
   def dataSource: DataSource[I, A]
   def identities: NonEmptyList[I]
 }
 
+trait FetchError extends Throwable with Product with Serializable
```

Comment:
I'm not sure FetchError is a good name, I'd call it FetchException as Error indicates potentially a real JVM Error when I read it and I would not dare to catch that.